File: app/src/main/java/com/worldventures/dreamtrips/modules/friends/service/command/GetSearchUsersCommand.java
package com.worldventures.dreamtrips.modules.friends.service.command;


import com.worldventures.dreamtrips.R;
import com.worldventures.dreamtrips.api.friends.SearchFriendsHttpAction;
import com.worldventures.dreamtrips.api.friends.model.ImmutableSearchParams;
import com.worldventures.dreamtrips.api.friends.model.SearchParams;
import com.worldventures.dreamtrips.modules.common.model.User;

import java.util.List;

import io.techery.janet.command.annotations.CommandAction;

@CommandAction
public class GetSearchUsersCommand extends GetUsersCommand {

   private String query;
   private int page;
   private int perPage;

   public GetSearchUsersCommand(String query, int page, int perPage) {
      this.query = query;
      this.page = page;
      this.perPage = perPage;
   }

   @Override
   protected void run(CommandCallback<List<User>> callback) throws Throwable {
      janet.createPipe(SearchFriendsHttpAction.class)
            .createObservableResult(new SearchFriendsHttpAction(provideSearchParams()))
            .map(SearchFriendsHttpAction::response)
            .map(this::convert)
            .subscribe(callback::onSuccess, callback::onFail);
   }

   private SearchParams provideSearchParams() {
      return ImmutableSearchParams.builder()
            .page(page)
            .perPage(perPage)
            .query(query)
            .build();
   }

   @Override
   public int getFallbackErrorMessage() {
      return R.string.error_failed_to_load_users;
   }
}

File: app/src/main/java/com/worldventures/dreamtrips/social/service/friends/interactor/command/GetSearchUsersCommand.kt
package com.worldventures.dreamtrips.social.service.friends.interactor.command

import com.worldventures.core.model.User
import com.worldventures.dreamtrips.R
import com.worldventures.dreamtrips.api.friends.SearchFriendsHttpAction
import com.worldventures.dreamtrips.api.friends.model.ImmutableSearchParams
import com.worldventures.dreamtrips.api.friends.model.SearchParams
import io.techery.janet.Command

import io.techery.janet.command.annotations.CommandAction

@CommandAction
class GetSearchUsersCommand(val query: String, val page: Int, val perPage: Int) : GetUsersCommand() {

   @Throws(Throwable::class)
   override fun run(callback: Command.CommandCallback<List<User>>) {
      janet.createPipe(SearchFriendsHttpAction::class.java)
            .createObservableResult(SearchFriendsHttpAction(provideSearchParams()))
            .map { it.response() }
            .map { this.convert(it) }
            .subscribe(callback::onSuccess, callback::onFail)
   }

   private fun provideSearchParams(): SearchParams {
      return ImmutableSearchParams.builder()
            .page(page)
            .perPage(perPage)
            .query(query)
            .build()
   }

   override fun getFallbackErrorMessage(): Int = R.string.error_failed_to_load_users
}
